sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlanDistinctKeys.scala

```diff
@@ -29,6 +29,12 @@ import org.apache.spark.sql.internal.SQLConf.PROPAGATE_DISTINCT_KEYS_ENABLED
  */
 trait LogicalPlanDistinctKeys { self: LogicalPlan =>
   lazy val distinctKeys: Set[ExpressionSet] = {
-    if (conf.getConf(PROPAGATE_DISTINCT_KEYS_ENABLED)) DistinctKeyVisitor.visit(self) else Set.empty
+    if (conf.getConf(PROPAGATE_DISTINCT_KEYS_ENABLED)) {
+      val keys = DistinctKeyVisitor.visit(self)
+      require(keys.forall(_.nonEmpty))
```

Comment:
Do we really need this `require`? It looks fine to have an empty set as the distinct keys, e.g. global aggregate without keys. It means that the entire data set is distinct (have at most one row), and `EliminateDistinct` is OK with empty set in distinct keys.

Reply:
I think it's more about avoid some unexpected things. It will be a correctness issue if other opterators return empty distinct key. And as you mentioned, the global aggregate has already optimzied by `EliminateDistinct`, so it's fine ?

Reply:
My point is that `DistinctKeyVisitor` does not work with global aggregate now, and an empty expression set is still a valid distinct key, why do we forbid it?

Reply:
We have already forbidden it inside `DistinctKeyVisitor`. Do you think we should support that case ?
https://github.com/apache/spark/blob/a67acbaa29d1ab9071910cac09323c2544d65303/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala#L50

Reply:
This is only done at the `else` branch, not the `if` branch. I think we have two options:
1. keep the requirement, and add the `filter` in the `if` branch as well
2. remove the requirement, and remove the `filter` from the `else` branch.

I prefer option 2 as I think an empty expression set does mean something as a distinct key, we should not ignore this information. It also works the same as other distinct keys:
1. It can replace all other distinct keys as it's a subset of any expression set
2. It can satisfy any distinct key requirement, e.g. remove unnecessary distinct in aggregate functions.